doc/autodiscover.7
.\" SPDX-License-Identifier: CC-BY-SA-4.0 or-later
.\" SPDX-FileCopyrightText: 2021-2024 grommunio GmbH
.TH autodiscover 7 "" "Gromox" "Gromox admin reference"
.SH Name
autodiscover \(em AutoDiscover protocols
.SH Description
.PP
AutoDiscover is a HTTP-based discovery protocol, originally introduced for
Exchange 2007, that helps users configure
their email client settings automatically. The MS-OXDSCLI specification
speaks about locating such AutoDiscover service, the MS-OXDISCO specification
speaks about the XML exchange between such a service and a client. These terms
are herein used to avoid ambiguity in the "AD" acronym (ActiveDirectory).
.PP
Autodiscover V2 is not an improved version, it is an extra layer that warrants
a disgruntled remark about questionable protocol design. Locating the
AutoDiscover server still happens via DNS or AD-SCP (Service Connection Points
in Active Directory) query. The V2 request
contains the user identity and the name of a next-level protocol that the
client seeks, e.g. "ActiveSync", "EWS" or "AutoDiscoverV1". The response is now
a JSON document and generally contains just one URL, namely for the service
sought. Indeed there is no way to obtain MAPI, IMAP or SMTP information in
Autodiscover V2.
.PP
Note that Autodiscover is not used exclusively by Microsoft Outlook,
Autodiscover is the main discovery protocol for any EAS-enabled device and
application, such as Apple iOS, Android and other applications.
.SH OXDSCLI summary
The MS-OXDSCLI document is about how clients can collect one or more
AutoDiscover URLs/endpoints that, in a second step (OXDISCO), they can make
HTTP POST requests to to retrieve server configuration. Behaviors that have
been specified in MS-OXDSCLI or have been observed in practice:
.IP \(bu 4
Using an URL from a previous successful run.
.IP \(bu 4
Performing a DNS TXT lookup on the domain name to check for the presence of a
Microsoft 365 account identifier and, if found, using the URL
<\fBhttps://outlook.office365.com/autodiscover/autodiscover.xml\fP>.
.IP \(bu 4
When joined to an NT Domain/ActiveDirectory, performing an LDAP lookup for
service connection points (SCP), i.e. objects matching
`(&(objectClass=serviceConnectionPoint)
(serviceClassName=ms-Exchange-AutoDiscover-Service))`, and using the URLs
obtained from "serviceBindingInformation" attributes.
.IP \(bu 4
"Root domain" method: Constructing a string based on the user's e-mail address
(e.g. u@example.com),
<\fBhttps://\fP\fIexample.com\fP\fB/autodiscover/autodiscover.xml\fP>.
.IP \(bu 4
"AutoDiscover domain" method: Constructing a string based on the user identity,
<\fBhttps://autodiscover.\fP\fIexample.com\fP\fB/autodiscover/autodiscover.xml\fP>.
.IP \(bu 4
Performing a DNS lookup of the name \fB_autodiscover._tcp.\fP\fIdomain\fP of
record type SRV, and from the response data, constructing the string
\fBhttps://\fP\fIrdata\fP[\fB:\fP\fIport\fP]\fB/autodiscover/autodiscover.xml\fP.
.IP \(bu 4
Utilizing an unspecified "local mechanism" method. (might be the same
as using lastrun URL?)
.IP \(bu 4
"Redirect" method: Performing an unauthenticated HTTP GET request to
<\fBhttp://autodiscover.\fP\fIdomain\fP\fB/autodiscover/autodiscover.xml\fP,
and upon receiving a HTTP 302 response, using the URL in the "Location:"
response header for AutoDiscover.
.PP
Gromox recommends that administrators employ the AutoDiscover Domain method,
i.e. add a \fBautodiscover.\fP\fIdomain.com\fP entry to their DNS zones. This
is beneficial over the root domain method, because the host serving up e.g. a
company's website need not be taught about autodiscover paths. For
\fBautodiscover.\fP\fIdomain.com\fP, either an AAAA/A pair \fBor\fP CNAME
resource record can be used. Split-horizon DNS systems \fBmust\fP publish the
autodiscover DNS record(s) in the publicly visible set. The AutoDiscover URL
must equally be reachable from any network segment where access is expected
from \(em if need be, set up HTTP reverse proxies.
.SH OXDISCO summary
The MS-OXDISCO document specifies what clients should do once they have an URL
collection obtained from the OXDSCLI mechanism(s).
.PP
When a user sets up a new email account or changes their existing email client
settings, the email client sends an authenticated HTTP POST request to an
Autodiscover URL and conveying the user's email address in the request body.
The server then responds with an XML document that contains the necessary
configuration settings.
.PP
The response contains information such as
.IP \(bu 4
the list of supported mail protocols and transports (e.g. MSRPC/RPCH/MAPIHTTP,
IMAP, SMTP, etc.)
.IP \(bu 4
the connection parameters for those (e.g. name of the home server, HTTP
endpoint URLs)
.IP \(bu 4
for MAPI, any extra mailboxes that should be opened unconditionally (e.g.
delegators, public folder)
.PP
The email client uses this information to configure the user's email account
automatically.
.SH gromox-dscli notes
The gromox\-dscli(8) utility can be used to diagnose problems with AutoDiscover
from a command line prompt.
.PP
gromox-dscli performs the following probes and in this order: RootDomain,
AutoDiscoverDomain, SRV.
.SH Outlook notes
When Outlook is running, there is an Outlook icon in the Windows taskbar's
notification area. By pressing Ctrl+RightMouseBtn, a service menu can be
brought up, which offers a "Test AutoDiscover" command for diagnosing problems
from Windows. Known bugs: The dialog may ignore the contents of the password
field and instead use a saved password or SSO, leading to potentially
unanticipated authentication successes or failures. If in doubt, use
gromox-dscli.
.PP
When an AutoDiscover response contains broken information (e.g. unreachable
endpoints due to a faulty oxdisco_exonym setting on the server), Outlook may
refuse to open mailboxes and the application may exit prematurely. AutoDiscover
responses are cached on disk under `%LOCALAPPDATA%/Microsoft/Outlook/* -
Autodiscover.xml` and/or `%LOCALAPPDATA%/Microsoft/Outlook/16/AutoD.*.xml` and
can be deleted/diagnosed as needed.
.PP
Outlook performs the following probes in this primary order: TXT, LDAP-SCP,
RootDomain, AutoDiscoverDomain, SRV, Local, Redirect. Secondary order that was
observed: TXT, LDAP-SCP, RootDomain, AutoDiscoverDomain, Local, Redirect, SRV.
.PP
Individual probes can be disabled via Windows Registry in
.IP \(bu 4
HKEY_CURRENT_USER\\Software\\Microsoft\\Office\\16.0\\Outlook\\AutoDiscover
.PP
and/or
.IP \(bu 4
HKEY_CURRENT_USER\\Software\\Policies\\Microsoft\\Office\\16.0\\Outlook\\AutoDiscover,
.PP
by setting one or more of
.IP \(bu 4
ExcludeLastKnownGoodURL=DWORD:1
.IP \(bu 4
ExcludeExplicitO365Endpoint=DWORD:1
.IP \(bu 4
EnableOffice365ConfigService=DWORD:0
.IP \(bu 4
ExcludeScpLookup=DWORD:1
.IP \(bu 4
ExcludeHttpsRootDomain=DWORD:1
.IP \(bu 4
ExcludeHttpsAutoDiscoverDomain=DWORD:1
.IP \(bu 4
ExcludeSrvRecord=DWORD:1
.IP \(bu 4
ExcludeHttpRedirect=DWORD:1
.PP
as desired. For the Group Policy editor, ADMX template and ADML
language packs are available from
<https://www.microsoft.com/en-us/download/details.aspx?id=49030>.
.PP
When a DNS zone is M365-enabled, Outlook opens a mini browser window for
authenticating with M365. To prevent this, you can set
ExcludeExplicitO365Endpoint=1 as described.
.PP
Outlook stops probing after the first successful AutoDiscover HTTP POST
request. A non-responsive AutoDiscovery server (firewall DROP policy, or TCP
RST) is treated the same as a 404 Not Found response.
.PP
Known bugs: Outlook ignores the port number in the DNS SRV response. Outlook
and/or the Windows HTTP libraries also erroneously show a warning popup
whenever the hostname in the SRV result does not match the e-mail domain (even
under MS Exchange). Redirection is the key idea of an SRV record and, as far as
security considerations go, is no more significant than following a CNAME-typed
autodiscover.example.com record.
.PP
Outlook re-runs AutoDiscover periodically in the background. This can cause
popups such as re-authentication or SRV warnings (particularly after a
temporary outage).
.SH Testing scenarios
To force using a particular Autodiscover server in Windows, such as when Gromox
is run in a development environment with a fake domain,
c:\\windows\\system32\\drivers\\etc\\hosts can be populated with a static entry
for \fBautodiscover.\fP\fIexample.com\fP to get that particular scenario
working.
.SH See also
\fBgromox\fP(7), \fBautoconfig\fP(7), \fBautodiscover\fP(4gx)


doc/changelog.rst
Gromox 3.8 (2026-06-30)
=======================

Enhancements:

* The Information Store now supports a per-store split-process model.
  The gromox.cfg:istore_standalone config directive has turned into a bitmask.
* emsmdb(Outlook): fnevObjectModified notifications for folders now convey
  the new counts of read/unread messages.
* Support for custom per-message keywords (a.k.a. color categories) in IMAP
* Partial IMAP4rev2 support (ENABLE, NAMESPACE, SELECT/EXAMINE, STATUS
  SIZE/DELETED, ESEARCH, (UID) MOVE, LIST-EXTENDED, SEARCHRES)
* Conversions between HTML and RTF now support and prefer the use of the
  external converter Pandoc, if installed.
* Conversions from HTML to plaintext now support the use of Pandoc as a
  fallback if chawan/w3m is not present.
* mbop: new "exaddrxlat" command
* ruleproc: implemented OP_FORWARD handling for TwoStep processor
* exmdb, zcore: new link_message RPCs for bulk linking, speeds up
  external search folder filling agents (e.g. grommunio-web)

Fixes:

* ab_tree: restore/implement time-based purging of the cache
* imap: fixed BODY[HEADER]/[TEXT]/[HEADER.FIELDS] returning NIL
* Server side rule and meeting processor: On reception of meeting updates, the
  prior calendar entry is replaced (rather than left untouched, which caused
  duplicates).
* ruleproc now ignores stale meeting updates (e.g. if an older update was held
  up in SMTP queues and was delivered late past another)
* imap: searching by Bcc field is now supported
* imap: STORE commands no longer fail when unrecognized flags are used
* midb: the SENTBEFORE/SENTON/SENTSINCE imap search keywords now test
  against the Date header rather than PR_LAST_MODIFICATION_TIME
* midb: the SENT* keywords now disregard timezone as mandated
* imap: FETCH/STORE/SEARCH no longer produce EXPUNGE replies
* ews: no longer loses the attendee list of an appointment when saved

Behavioral changes:

* The default value for exmdb_provider.cfg:cache_interval has been reduced from
  15min to 1min, closing sqlite handles earlier when they are idle.
* exmdb_client: back off exponentially on notify reconnect
* ews: the SyncFolderItems and GetItem response sizes are now capped for
  reasons of memory use (clients should paginate as needed)
* ews: a limit is imposed on the streaming event backlog size now
  (clients should reconnect/resync as needed if they get kicked)


Gromox 3.7 (2026-04-30)
=======================

Fixes:

* Outlook no longer loses appointsments when a shared calendar is opened
* The "percentage complete" of an appointment was mangled when passing through
  EWS, which has been fixed. (EWS uses [0, 100], but MAPI uses [0, 1.0])
* EWS now synthesizes a value for the "Received Time" field.
* AutoDiscover no longer erroneously advertises OWA API support.
  This had made it impossible for Thunderbird to connect.
* EWS: Setting the read flag on a message no longer generates the
  E-2901/E-3418 error message on the server.
* The EWS implementation now performs retrieval/update of Out Of Office
  settings such that it works in multi-server environments.
* The use of Qmail as an outgoing SMTP server is now supported by the message
  sending routines.
* Fix a case of Outlook notifications getting potentially suppressed.
* Fix a case of `sqlite_busy on /var/lib/gromox/user/xyz: held by unknown`
  log message appearing in the exmdb daemon (gromox-http/istore).
* Conversion from MAPI to iCal appoinments used the wrong UID for exceptions
  in a recurring series, which has been fixed.
* Reading iCal timezone definitions mangled the daylightbias, which was fixed.
* Logins with Outlook 2010 now work when Gromox is running with OpenSSL 3.
  (OpenSSL 1.1 & LibreSSL were fine, but OSSL3 changed behavior.)
* SMIME messages signed multiple times are unwrapped recursively
  and will now show the body. (Applies to just g-web; OL worked previously.)

Enhancements:

* Added conversion routines to zcore so that g-web can export GAL items to vCard


Gromox 3.6 (2026-03-26)
=======================

Enhancements:

* Gromox has gained a component that now serves the Offline Addressbook.
* EWS: Task operations (creation, update, property retrieval) have been
  implemented.
* EWS: Calendar occurrence operations (create, update, delete), including
  timezone support for those, have been implemented.
* EWS: A number of request have been implemented: ExpandDL, GetPersona,
  CreateUserConfiguration, UpdateUserConfiguration, DeleteUserConfiguration,
  DeleteAttachment, AddDelegate, UpdateDelegate and RemoveDelegate.
* EWS: Lifecycle for meetings have been implemented, i.e. invitation sending
  when the `UpdateItem` operation is exercised, cancellation sending on
  `DeleteItem`, and response sending to the organizer when a participant
  accepts/declines a request.
* EWS: Read receipts are now generated when a user messages are marked as read;
  the `SuppressReadReceipts` attribute of the `UpdateItem` operation is honored
  as well.
* EWS: The `ResolveNames` operation now performs partial matches too.
* EWS: Added delegate permission level support and cross-mailbox permission
  checks.
* EWS: eM Client reported "no conversion for Message::Flag" when toggling the
  flagged state of a message, which has been fixed/implemented.
* The server-side meeting autoprocessor now updates the organizer's calendar
  when a meeting response is received, and sets the flag that XY has responded.
* The AutoDiscover handler now supports the impersonation username syntax
  (auth@a.de!sharedmbox@b.de) in conjunction with EAS
* Conversion of appointments from MAPI to iCal form now can deal with a
  zero-length PidLidAppointmentTimeZoneDefinitionStartDisplay attribute
* The eml2mt conversion utility now emits messages sooner (previously it
  collected all messages first before outputting)
* Conversions from HTML to plaintext now support and prefer the use of the
  external converter Chawan (if installed)

Fixes:


exch/php/ews/autodiscover.php
<?php

require_once "../lib/db.php";
require_once "../lib/util.php";

define("RESPONSE_XMLNS", "http://schemas.microsoft.com/exchange/autodiscover/responseschema/2006");
define("RESPONSE_OUTLOOK_XMLNS", "http://schemas.microsoft.com/exchange/autodiscover/outlook/responseschema/2006a");
define("SERVER_VERSION", "73C0834F"); //15.00.0847.4040
define("RESPONSE_MOBILE_XMLNS", "http://schemas.microsoft.com/exchange/autodiscover/mobilesync/responseschema/2006");


/*
request list like below:

<Autodiscover xmlns="http://schemas.microsoft.com/exchange/autodiscover/outlook/requestschema/2006">
	<Request>
		<EMailAddress>user@contoso.com</EMailAddress>
		<LegacyDN>/o=contoso/ou=Exchange Administrative Group (FYDIBOHF23SPDLT)/cn=Recipients/cn=user
		<AcceptableResponseSchema>http://schemas.microsoft.com/exchange/autodiscover/outlook/responseschema/2006a</AcceptableResponseSchema>
	</Request>
</Autodiscover>
*/

if (!isset($_SERVER['HTTPS'])) {
	require_once "../lib/conf.php";
	header("Location: https://" . get_default_hostname() . "/Autodiscover/Autodiscover.xml");
	exit;
}

if (0 != strcasecmp($_SERVER['REQUEST_METHOD'], "POST")) {
	die("E-2000: invalid request method, must be POST!");
}

$xml_in = file_get_contents("php://input");
error_log("Autodiscover Request:\n" . $xml_in, 0);
$xml_request = simplexml_load_string($xml_in);
if ($xml_request === false ||
    strcasecmp($xml_request->getName(), "Autodiscover") != 0)
	die("E-2001: invalid request xml root, should be Autodiscover");
$email_address = $xml_request->Request->EMailAddress;
if (!isset($email_address)) {
	die("E-2002: cannot find Request/EMailAddress in xml request");
}
if (!strpos($email_address, '@')) {
	die("E-2003: format error with Request/EMailAddress in xml request");
}

if ('public.folder.root' == substr($email_address, 0, strpos($email_address, "@"))) {
	$domain = substr($email_address, strpos($email_address, "@") + 1);
	$dinfo = get_domain_info_by_name($domain);
	if (!$dinfo) {
		die("E-2004: cannot find domain information");
	}
	$Autodiscover = new SimpleXMLElement('<?xml version="1.0" encoding="utf-8"?><Autodiscover></Autodiscover>');
	$Autodiscover->addAttribute('xmlns', RESPONSE_XMLNS);
	$Response = $Autodiscover->addChild('Response');
	$Response->addAttribute('xmlns', RESPONSE_OUTLOOK_XMLNS);
	$User = $Response->addChild('User');
	$User->addChild('AutoDiscoverSMTPAddress', 'public.folder.root@' . $domain);
	$User->addChild('DisplayName', 'Public Folder');
	$User->addChild('LegacyDN', publicfolder_to_essdn($dinfo));
	$User->addChild('DeploymentId', get_domain_server_guid($dinfo));
	$Account = $Response->addChild('Account');
	$Account->addChild('AccountType', 'email');
	$Account->addChild('Action', 'settings');
	$host_name = get_default_hostname();
	$server_name = get_domain_server_guid($dinfo) . '@' . $dinfo['domain'];
	if (advertise_mh($_SERVER["HTTP_USER_AGENT"])) {
		$Protocol = $Account->addChild('Protocol');
		$Protocol->addChild('Type', 'EXHTTP');
		$Protocol->addChild('Server', get_http_proxy($dinfo['homedir'], $host_name));
		$Protocol->addChild('SSL', 'On');
		$Protocol->addChild('CertPrincipalName', 'None');
		$Protocol->addChild('AuthPackage', 'basic');
		$Protocol->addChild('ServerExclusiveConnect', 'on');
		$Protocol = $Account->addChild('Protocol');
		$Protocol->addAttribute('Type', 'mapiHttp');
		$Protocol->addAttribute('Version', '1');
		$MailStore = $Protocol->addChild('MailStore');
		$MailStore->addChild('InternalUrl', 'https://' . get_http_proxy($dinfo['homedir'], $host_name) . "/mapi/emsmdb/?MailboxId=" . $server_name);
		$MailStore->addChild('ExternalUrl', 'https://' . get_http_proxy($dinfo['homedir'], $host_name) . "/mapi/emsmdb/?MailboxId=" . $server_name);
		$AddressBook = $Protocol->addChild('AddressBook');
		$AddressBook->addChild('InternalUrl', 'https://' . get_http_proxy($dinfo['homedir'], $host_name) ."/mapi/nspi/?MailboxId=" . $server_name);
		$AddressBook->addChild('ExternalUrl', 'https://' . get_http_proxy($dinfo['homedir'], $host_name) ."/mapi/nspi/?MailboxId=" . $server_name);
	}
	if (advertise_rpch($_SERVER["HTTP_USER_AGENT"])) {
		$Protocol = $Account->addChild('Protocol');
		$Protocol->addChild('Type', 'EXCH');
		$Protocol->addChild('Server', $server_name);
		$Protocol->addChild('ServerVersion', SERVER_VERSION);
		$Protocol->addChild('ServerDN', get_server_dn($server_name));
		$Protocol->addChild('MdbDN', get_mdb_dn($server_name));
		if (0 == strncasecmp($_SERVER['HTTP_USER_AGENT'], "Microsoft Office/", 17)
			&& floatval(substr($_SERVER['HTTP_USER_AGENT'], 17)) >= 15) {
			$Protocol->addChild('AuthPackage', 'anonymous');
		} else {
			$Protocol->addChild('AuthPackage', 'ntlm');
		}
		$Protocol->addChild('ServerExclusiveConnect', 'off');
		$Protocol = $Account->addChild('Protocol');
		$Protocol->addChild('Type', 'EXPR');
		$Protocol->addChild('Server', get_http_proxy($dinfo['homedir'], $host_name));
		$Protocol->addChild('SSL', 'On');
		$Protocol->addChild('CertPrincipalName', 'None');
		$Protocol->addChild('AuthPackage', 'basic');
		$Protocol->addChild('ServerExclusiveConnect', 'on');
	}
} else {
	// Private mbox
	if (!isset($_SERVER['REMOTE_USER'])) {
		header("Status: 401 Unauthorized");
		header("Content-Length: 0");
		header("WWW-Authenticate: Basic realm=" . $_SERVER['SERVER_NAME']);
		exit;
	}

	if (isset($xml_request->Request->AcceptableResponseSchema) &&
		0 == strcasecmp($xml_request->Request->AcceptableResponseSchema,
		RESPONSE_MOBILE_XMLNS)) {
		/* ActiveSync Autodiscover */
		$uinfo = get_user_info_by_name($email_address);
		if (!$uinfo) {
			die("E-2005: cannot find email address information");
		}

		$Autodiscover = new SimpleXMLElement('<?xml version="1.0" encoding="utf-8"?><Autodiscover></Autodiscover>');
		$Autodiscover->addAttribute('xmlns', RESPONSE_XMLNS);
		$Response = $Autodiscover->addChild('Response');
		$Response->addAttribute('xmlns', RESPONSE_MOBILE_XMLNS);
		$Response->addChild('Culture', 'en:us');
		$User = $Response->addChild('User');
		$host_name = get_default_hostname();
		$server_url = "https://" . get_http_proxy($uinfo['maildir'], $host_name) . "/Microsoft-Server-ActiveSync";
		$User->addChild('DisplayName', htmlspecialchars($uinfo['real_name']));
		$User->addChild('EMailAddress', $uinfo['username']);
		$Action = $Response->addChild('Action');
		$Settings = $Action->addChild('Settings');
		$Server = $Settings->addChild('Server');
		$Server->addChild('Type', 'MobileSync');
		$Server->addChild('Url', $server_url);
		$Server->addChild('Name', $server_url);

	} else {
		$legacy_dn = $xml_request->Request->LegacyDN;
		if ($legacy_dn) {
			$uinfo = essdn_to_username($legacy_dn);
			if (!$uinfo) {
				die("E-2006: cannot find essdn information");
			}
		} else {
			$uinfo = get_user_info_by_name($email_address);
			if (!$uinfo) {
				die("E-2007: could not resolve $email_address (do you have the right database set?)");
			}
		}

		$Autodiscover = new SimpleXMLElement('<?xml version="1.0" encoding="utf-8"?><Autodiscover></Autodiscover>');
		$Autodiscover->addAttribute('xmlns', RESPONSE_XMLNS);
		$Response = $Autodiscover->addChild('Response');
		$Response->addAttribute('xmlns', RESPONSE_OUTLOOK_XMLNS);
		$User = $Response->addChild('User');
		$User->addChild('AutoDiscoverSMTPAddress', $uinfo['username']);
		$User->addChild('DisplayName', htmlspecialchars($uinfo['real_name']));
		$User->addChild('LegacyDN', username_to_essdn($uinfo));
		$User->addChild('EMailAddress', $uinfo['username']);
		$User->addChild('DeploymentId', get_user_server_guid($uinfo));
		$Account = $Response->addChild('Account');
		$Account->addChild('AccountType', 'email');
		$Account->addChild('Action', 'settings');
		$host_name = get_default_hostname();
		$server_name = get_user_server_guid($uinfo) . '@' . $uinfo['domain'];
		if (advertise_mh($_SERVER["HTTP_USER_AGENT"])) {
			$Protocol = $Account->addChild('Protocol');
			$Protocol->addChild('OOFUrl', 'https://' . $host_name . '/EWS/Exchange.asmx');
			$Protocol->addChild('OABUrl', 'https://' . $host_name . '/OAB/');
			$Protocol->addChild('Type', 'EXHTTP');
			$Protocol->addChild('Server', get_http_proxy($uinfo['maildir'], $host_name));
			$Protocol->addChild('SSL', 'On');
			$Protocol->addChild('CertPrincipalName', 'None');
			$Protocol->addChild('AuthPackage', 'basic');
			$Protocol->addChild('ASUrl', 'https://' . $host_name . '/EWS/Exchange.asmx');
			$Protocol->addChild('EwsUrl', 'https://' . $host_name . '/EWS/Exchange.asmx');
			$Protocol->addChild('EmwsUrl', 'https://' . $host_name . '/EWS/Exchange.asmx');
			$Protocol->addChild('EcpUrl', 'https://' . $host_name . '/EWS/');
			$Protocol->addChild('EcpUrl-photo', 'thumbnail.php');
			$Protocol->addChild('ServerExclusiveConnect', 'on');
			$Protocol = $Account->addChild('Protocol');
			$Protocol->addAttribute('Type', 'mapiHttp');
			$Protocol->addAttribute('Version', '1');
			$MailStore = $Protocol->addChild('MailStore');
			$MailStore->addChild('InternalUrl', 'https://' . get_http_proxy($uinfo['maildir'], $host_name) . "/mapi/emsmdb/?MailboxId=" . $server_name);
			$MailStore->addChild('ExternalUrl', 'https://' . get_http_proxy($uinfo['maildir'], $host_name) . "/mapi/emsmdb/?MailboxId=" . $server_name);
			$AddressBook = $Protocol->addChild('AddressBook');
			$AddressBook->addChild('InternalUrl', 'https://' . get_http_proxy($uinfo['maildir'], $host_name) ."/mapi/nspi/?MailboxId=" . $server_name);
			$AddressBook->addChild('ExternalUrl', 'https://' . get_http_proxy($uinfo['maildir'], $host_name) ."/mapi/nspi/?MailboxId=" . $server_name);
		}
		if (advertise_rpch($_SERVER["HTTP_USER_AGENT"])) {
			$Protocol = $Account->addChild('Protocol');
			$Protocol->addChild('OOFUrl', 'https://' . $host_name . '/EWS/Exchange.asmx');
			$Protocol->addChild('OABUrl', 'https://' . $host_name . '/OAB/');
			$Protocol->addChild('Type', 'EXCH');
			$Protocol->addChild('Server', $server_name);
			$Protocol->addChild('ServerVersion', SERVER_VERSION);
			$Protocol->addChild('ServerDN', get_server_dn($server_name));
			$Protocol->addChild('MdbDN', get_mdb_dn($server_name));
			if (0 == strncasecmp($_SERVER['HTTP_USER_AGENT'], "Microsoft Office/", 17)
				&& floatval(substr($_SERVER['HTTP_USER_AGENT'], 17)) >= 15) {
				$Protocol->addChild('AuthPackage', 'anonymous');
			} else {
				$Protocol->addChild('AuthPackage', 'ntlm');
			}
			$Protocol->addChild('PublicFolderServer', get_default_hostname());
			$Protocol->addChild('ASUrl', 'https://' . $host_name . '/EWS/Exchange.asmx');
			$Protocol->addChild('EwsUrl', 'https://' . $host_name . '/EWS/Exchange.asmx');
			$Protocol->addChild('EmwsUrl', 'https://' . $host_name . '/EWS/Exchange.asmx');
			$Protocol->addChild('EcpUrl', 'https://' . $host_name . '/EWS/');
			$Protocol->addChild('EcpUrl-photo', 'thumbnail.php');
			$Protocol->addChild('ServerExclusiveConnect', 'off');
			$Protocol = $Account->addChild('Protocol');
			$Protocol->addChild('Type', 'EXPR');
			$Protocol->addChild('Server', get_http_proxy($uinfo['maildir'], $host_name));
			$Protocol->addChild('SSL', 'On');
			$Protocol->addChild('CertPrincipalName', 'None');
			$Protocol->addChild('AuthPackage', 'basic');
			$Protocol->addChild('ServerExclusiveConnect', 'on');
			$Protocol->addChild('OOFUrl', 'https://' . $host_name . '/EWS/Exchange.asmx');
			$Protocol->addChild('OABUrl', 'https://' . $host_name . '/OAB/');
		}

		$altMailboxes = get_secondary_store_hints($email_address);
		foreach ($altMailboxes as $altMailbox) {
			$AlternativeMailbox = $Account->addChild('AlternativeMailbox');
			$AlternativeMailbox->addChild('Type', 'Delegate');
			$AlternativeMailbox->addChild('DisplayName', htmlspecialchars($altMailbox['DisplayName']));
			$AlternativeMailbox->addChild('SmtpAddress', $altMailbox['SmtpAddress']);
			$AlternativeMailbox->addChild('OwnerSmtpAddress', $altMailbox['SmtpAddress']);
		}

		$publicfolder = $Account->addChild('PublicFolderInformation');
		$publicfolder->addChild('SmtpAddress', 'public.folder.root' . substr($email_address, strpos($email_address, "@")));
	}
}
error_log("Autodiscover Response:\n" . $Autodiscover->asXML(), 0);

header("Content-Type: text/xml; charset=utf-8");
print $Autodiscover->asXML();

exit;

?>
